### ISSUE TITLE
env NODE_EXTRA_CA_CERTS  not work

### ISSUE THREAD

**Issue by magicode** [electron/electron#10257]
* Electron version: 1.6.7
* Operating system: win7

### Expected behavior

env NODE_EXTRA_CA_CERTS  should work
node doc  https://nodejs.org/dist/latest-v6.x/docs/api/cli.html#cli_node_extra_ca_certs_file

### Actual behavior

env NODE_EXTRA_CA_CERTS  not work

### How to reproduce

electron no run this line
https://github.com/electron/node/blob/dc8fe9d390b1deb32ae2601357a90679cbcc757e/src/node.cc#L4659

**Comment by bontibon** [comment]
Just ran into this issue as well.  I was hoping I could get around it by locally trusting my CA cert and using `NODE_OPTIONS=--use-openssl-ca`, but that didn't work either.

**Comment by nerakhon** [comment]
The issue is reproducible on Linux and macOS as well. Just adding for completeness. There is a workaround by using node extra ca libraries but this core node functionality should IMO work.

**Comment by redders6600** [comment]
@nerakhon - could you provide a link to the libraries that you're using as a workaround?

**Comment by nerakhon** [comment]
@redders6600 ssl-root-cas : https://www.npmjs.com/package/ssl-root-cas

**Comment by magicode** [comment]
@nerakhon @redders6600 

if you write the software you can add this code to support NODE_EXTRA_CA_CERTS env.

```js
(function(){
	if(!process.env.NODE_EXTRA_CA_CERTS) return;
	try{
		var netfreeCa = require("fs").readFileSync(process.env.NODE_EXTRA_CA_CERTS);
	}catch(e){
		return;
	}
	
	var NativeSecureContext = process.binding('crypto').SecureContext;
	var oldaddRootCerts = NativeSecureContext.prototype.addRootCerts;
	NativeSecureContext.prototype.addRootCerts = function(){
		var ret = oldaddRootCerts.apply(this,arguments);
		this.addCACert(netfreeCa);
		return ret;
	};
})();
```

but you cannot add cacerts to other software based electron like https://desktop.github.com/

**Comment by redders6600** [comment]
@magicode thanks for the pointer, but this doesn't seem to work for me - the function gets run as the first part of my electron app code, but the new `addRootCerts` function is never called...

**Comment by magicode** [comment]
maybe have error in read file NODE_EXTRA_CA_CERTS

**Comment by redders6600** [comment]
My `proxyRules` were wrong... I was blinkered and thought it must be the certs, but the bug was elsewhere! This is working well, thanks for your help :+1: .

Looking forward to removing this code once this issue is resolved.

**Comment by nikitakit** [comment]
I've run into the same issue with Atom on macOS.

I develop an Atom package that involves making requests to https servers, including servers that use self-signed certificates. I would like to recommend that users with self-signed certificates take advantage of built-in node capabilities to authorize them, for example NODE_EXTRA_CA_CERTS. It is unfortunate that electron does not currently support this setting. Currently the easiest alternative is to set NODE_TLS_REJECT_UNAUTHORIZED, which is not ideal because it disables security checks throughout the entire application.

**Comment by margolisj** [comment]
@magicode @redders6600 Is there any where specific I need to place this code? Is there a way confirm that the extra cert was correctly added? I know its executing but I'm not sure if electron is picking up the new cert. Thanks.

**Comment by codebytere** [comment]
We are no longer implementing bugfixes for versions of Electron <= `1.7.x`, so i'm going to close this issue but if it is still persisting in more recent versions of Electron we can certainly reopen it!

**Comment by codenirvana** [comment]
@codebytere Same issue with Electron `v4.0.5`.
Is there any way to debug this? `process.env.NODE_EXTRA_CA_CERTS` is set correctly but its not extending the CA.

**Comment by Nuss9940** [comment]
Still a problem with `v5.0.7` on Arch Linux

**Comment by codebytere** [comment]
@Nuss9940 that's likely owing to [this](https://nodejs.org/api/cli.html#cli_node_options_options):
> This environment variable is ignored when node runs as setuid root or has Linux file capabilities set.

**Comment by nikitakit** [comment]
@codebytere I think this issue remains unfixed, and closing it is not appropriate without a suitable solution.

The underlying concern here is that Electron does not provide an option to change to extend the default list of trusted certificates/CAs with your own. This is useful when you would like to have an Electron app seamlessly interact with sites that have a self-signed certificate, or a corporate-issued certificate that's not in the default chain bundled with Electron.

The `NODE_OPTIONS` documentation for Electron explicitly says that `--use-bundled-ca`, `--use-openssl-ca`, and `--openssl-config` are unsupported, which are the only options that are relevant to this issue. I don't see how how `NODE_OPTIONS` applies with such limitations in place.

>This environment variable is ignored when node runs as setuid root or has Linux file capabilities set.
I don't see how this is relevant, either. The Atom text editor runs without setuid, and I'm on Mac so Linux file capabilities shouldn't be part of the picture either.

The real cause is that Electron does not run the normal `Start` method from node. Instead, it has its own copy that fails to respect the `NODE_EXTRA_CA_CERTS` variable.

In short, as far as I can tell the issue of being unable to modify the certificate list remains unchanged from two years ago, when this was first reported.

P.S. Actually, I lied that there's no way to override cert settings in Electron -- `NODE_TLS_REJECT_UNAUTHORIZED` works quite well for this purpose. I hope everyone can see why entirely disabling certificate checking is not an acceptable solution.

**Comment by vaskokj** [comment]
@nikitakit 

I would agree with you. I'm having the same exact issue with an app built upon electron 3.1.10. I've futzed around with `NODE_EXTRA_CA_CERTS`, the various `NODE_OPTIONS` (and then ran into this thread saying `--use-bundled-ca`, `--use-openssl-ca`, and `--openssl-config` aren't supported).

**Comment by danydev** [comment]
@codebytere Do you mind to reopen this? Looks like there are no working workarounds for this.

**Comment by jamischarles** [comment]
Here's how I got it working with the Loom Electron app:

1 Unpacked the electron app source with `asar`.
2. Added code to the source to listen for `NODE_EXTRA_CA_CERTS`:
(in my case I added it to the top of the `main.js` file

Code from
https://github.com/electron/electron/issues/10257#issuecomment-346310333
```js
(function(){
	if(!process.env.NODE_EXTRA_CA_CERTS) return;
	try{
		var extraca = require("fs").readFileSync(process.env.NODE_EXTRA_CA_CERTS);
	}catch(e){
		return;
	}
	
	var NativeSecureContext = process.binding('crypto').SecureContext;
	var oldaddRootCerts = NativeSecureContext.prototype.addRootCerts;
	NativeSecureContext.prototype.addRootCerts = function(){
		var ret = oldaddRootCerts.apply(this,arguments);
		this.addCACert(extraca);
		return ret;
	};
})();
```
3. Repacked it with `asar`. Replaced `loom.app/contents/Resources/app.asar` with my own packaged version.
4. Ran the binary from the terminal with
`env NODE_EXTRA_CA_CERTS='/usr/local/etc/openssl/certs/my_companys_custom_cert.pem'  ./loom.app/contents/MacOS/Loom`
5. no more cert issues in the electron process!!!!! hahahahah I can’t believe it

**Comment by SilverBut** [comment]
Still not working and your workarounds does not work on my side. @codebytere could you please kindly reopen this? The issue presents on Linux and macOS as well as on Windows.

**Comment by alex1701c** [comment]
> (node:26268) [DEP0111] DeprecationWarning: Access to process.binding('crypto') is deprecated.

Is there a workaround using non-deprecated API for accomplishing this?

**Comment by BCaspari** [comment]
> > (node:26268) [DEP0111] DeprecationWarning: Access to process.binding('crypto') is deprecated.
> 
> Is there a workaround using non-deprecated API for accomplishing this?

This worked for me:

`
import tls from "tls";
import fs from "fs";
import https from "https";

if (process.env.NODE_EXTRA_CA_CERTS && fs.existsSync(process.env.NODE_EXTRA_CA_CERTS)) {
    const cas = [...tls.rootCertificates];
    const buf = fs.readFileSync(process.env.NODE_EXTRA_CA_CERTS, "ascii");
    cas.push(buf);
    https.globalAgent.options.ca = cas;
}
`
